src/mds/CDir.cc

```diff
@@ -1260,15 +1262,60 @@ void CDir::assimilate_dirty_rstat_inodes_finish(MutationRef& mut, EMetaBlob *blo
     in->clear_dirty_rstat();
     blob->add_primary_dentry(dn, in, true);
   }
+}
+
+void CDir::add_dirty_rstat_inode(CInode *in)
+{
+  dirty_rstat_inodes.push_back(&in->item_dirty_rstat);
 
-  if (!dirty_rstat_inodes.empty())
-    inode->mdcache->mds->locker->mark_updated_scatterlock(&inode->nestlock);
+  auto pi = in->get_projected_inode();
+  if (pi->rstat.dirty_from.is_zero())
+    dout(1) << __func__ << " zero dirty_from " << *in << dendl;
+
+  if (rstat_dirty_from.is_zero() ||
+      rstat_dirty_from > pi->rstat.dirty_from) {
+    rstat_dirty_from = pi->rstat.dirty_from;
+    cache->mds->locker->mark_dirty_rstat_dirfrag(this);
+  }
+}
+
+void CDir::remove_dirty_rstat_inode(CInode *in)
+{
+  in->item_dirty_rstat.remove_myself();
+  if (dirty_rstat_inodes.empty() &&
+      (!is_auth() || get_projected_fnode()->rstat.dirty_from.is_zero())) {
+    item_dirty_rstat.remove_myself();
+    rstat_dirty_from = utime_t();
+  }
 }
 
 void CDir::mark_dirty_rstat(LogSegment *ls)
 {
   cache->mds->locker->mark_updated_scatterlock(&inode->nestlock);
   ls->dirty_dirfrag_nest.push_back(&inode->item_dirty_dirfrag_nest);
+
+  auto pf = get_projected_fnode();
+  if (pf->rstat.dirty_from.is_zero() &&
+      (pf->rstat.same_sums(pf->accounted_rstat) ||
+       pf->rstat.version != inode->get_projected_inode()->rstat.version))
+    return;
+
+  if (pf->rstat.dirty_from.is_zero())
+    dout(1) << __func__ << " zero dirty_from " << *this << dendl;
```

Comment:
nit: we probably don't want debug level 1 here

Reply:
"rstat dirty and dirty_from is zero" is unexpected. should only happens when upgrade from old mds